cpp/src/fil/infer.cu

```diff
@@ -452,6 +459,126 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_MANY_CLASSES> {
+      vector_leaf(vector_leaf) {
+    // Assign workspace
+    char* ptr = (char*)accumulate_workspace;
+    per_class_margin = (vec<NITEMS, float>*)ptr;
+    ptr += sizeof(vec<NITEMS, float>) * num_classes * num_threads_per_class;
+    vector_leaf_indices = (vec<NITEMS, int>*)ptr;
+    ptr += sizeof(vec<NITEMS, int>) * blockDim.x;
+    thread_num_rows = (int*)ptr;
+
+    // Initialise shared memory
+    for (int i = threadIdx.x; i < num_classes * num_threads_per_class;
+         i += blockDim.x) {
+      per_class_margin[i] = vec<NITEMS, float>();
+    }
+    vector_leaf_indices[threadIdx.x] = vec<NITEMS, int>();
+    thread_num_rows[threadIdx.x] = 0;
+    // __syncthreads() is called in infer_k
+  }
+  __device__ __forceinline__ void accumulate(
+    vec<NITEMS, int> single_tree_prediction, int tree, int num_rows) {
+    // Perform a transpose in shared memory
+    // Assign each thread to a class, so they can accumulate without atomics
+    __syncthreads();
+    // Write indices to shared memory
+    vector_leaf_indices[threadIdx.x] = single_tree_prediction;
+    thread_num_rows[threadIdx.x] = num_rows;
+    __syncthreads();
+    for (int i = threadIdx.x; i < num_classes * num_threads_per_class;
+         i += blockDim.x) {
+      int c = i % num_classes;
+      for (int j = threadIdx.x / num_classes; j < blockDim.x;
+           j += num_threads_per_class) {
+        for (int item = 0; item < thread_num_rows[j]; ++item) {
+          float pred =
+            vector_leaf[vector_leaf_indices[j][item] * num_classes + c];
+          per_class_margin[i][item] += pred;
+        }
+      }
+    }
+  }
+  // class probabilities or regression. for regression, num_classes
+  // is just the number of outputs for each data instance
+  __device__ __forceinline__ void finalize_multiple_outputs(float* out,
+                                                            int num_rows,
+                                                            int num_trees) {
+    __syncthreads();
+    for (int c = threadIdx.x; c < num_classes; c += blockDim.x) {
+#pragma unroll
+      for (int row = 0; row < num_rows; ++row) {
+        out[row * num_classes + c] = per_class_margin[c][row];
+      }
+    }
+  }
+  // using this when predicting a single class label, as opposed to sparse class vector
+  // or class probabilities or regression
+  __device__ __forceinline__ void finalize_class_label(float* out,
+                                                       int num_rows) {
+    __syncthreads();
+    int item = threadIdx.x;
+    int row = item;
+    if (item < NITEMS && row < num_rows) {
+      float max_margin = 0;
+      int best_class = 0;
+      for (int c = 0; c < num_classes; ++c) {
+        float margin = per_class_margin[c][row];
+        if (margin > max_margin) {
+          max_margin = margin;
+          best_class = c;
+        }
+      }
+      out[row] = best_class;
+    }
+  }
+  __device__ __forceinline__ void finalize(float* out, int num_rows,
```

Comment:
wouldn't the whole function be replaceable with GROVE_PER_CLASS_* finalize? You could even factor it out into a separate function to call from all 3 places.
Granted, that gives extra softmax support for probabilities, but you can always short-circuit it during model import if skl would never use it in multiclass RF

```
    acc = multi_sum<6>(per_thread, num_classes, blockDim.x / num_classes);
    if (threadIdx.x < num_classes) per_thread[threadIdx.x] = acc;
    __syncthreads();  // per_thread needs to be fully populated

    class_margins_to_global_memory(per_thread, per_thread + num_classes,
                                   transform, num_trees / num_classes,
                                   tmp_storage, out, num_rows, num_outputs);
```

Reply:
I used `multi_sum` where num_classes < blockDim.x and applied class_margins_to_global_memory.

Reply:
nice, thanks!